ext/v8/property-callback.cc

```diff
@@ -74,4 +117,112 @@ namespace rr {
+
+    VALUE rb_property = String(isolate, property->ToString());
+
+    Unlocker unlock(info.GetIsolate());
+    rb_funcall(code, rb_intern("call"), 2, rb_property, (VALUE)PropertyCallbackInfo::Integer(info));
+  }
+
+  void PropertyCallback::invokeNamedPropertyDeleter(v8::Local<v8::String> property, const v8::PropertyCallbackInfo<v8::Boolean>& info) {
+    v8::Isolate* isolate = info.GetIsolate();
+
+    VALUE code(unwrapDeleter(isolate, info.Data()));
+
+    VALUE rb_property = String(isolate, property->ToString());
+
+    Unlocker unlock(info.GetIsolate());
+    rb_funcall(code, rb_intern("call"), 2, rb_property, (VALUE)PropertyCallbackInfo::Boolean(info));
+  }
+
+  void PropertyCallback::invokeNamedPropertyEnumerator(const v8::PropertyCallbackInfo<v8::Array>& info) {
+    v8::Isolate* isolate = info.GetIsolate();
+
+    VALUE code(unwrapEnumerator(isolate, info.Data()));
+
+    Unlocker unlock(info.GetIsolate());
+    rb_funcall(code, rb_intern("call"), 1, (VALUE)PropertyCallbackInfo::Array(info));
+  }
+
+  void PropertyCallback::invokeIndexedPropertyGetter(uint32_t index, const v8::PropertyCallbackInfo<v8::Value>& info) {
+    v8::Isolate* isolate = info.GetIsolate();
+
+    VALUE code(unwrapGetter(isolate, info.Data()));
+
+    Unlocker unlock(info.GetIsolate());
+    rb_funcall(code, rb_intern("call"), 2, UINT2NUM(index), (VALUE)PropertyCallbackInfo::Value(info));
+  }
+
+  void PropertyCallback::invokeIndexedPropertySetter(uint32_t index, v8::Local<v8::Value> value, const v8::PropertyCallbackInfo<v8::Value>& info) {
+    v8::Isolate* isolate = info.GetIsolate();
+
+    VALUE code(unwrapSetter(isolate, info.Data()));
+
+    Unlocker unlock(info.GetIsolate());
+    rb_funcall(
+      code, rb_intern("call"), 3,
+      UINT2NUM(index),
+      (VALUE)Value(isolate, value),
+      (VALUE)PropertyCallbackInfo::Value(info)
+    );
+  }
+
+  void PropertyCallback::invokeIndexedPropertyQuery(uint32_t index, const v8::PropertyCallbackInfo<v8::Integer>& info) {
+    v8::Isolate* isolate = info.GetIsolate();
+
+    VALUE code(unwrapQuery(isolate, info.Data()));
+
+    Unlocker unlock(info.GetIsolate());
+    rb_funcall(code, rb_intern("call"), 2, UINT2NUM(index), (VALUE)PropertyCallbackInfo::Integer(info));
+  }
+
+  void PropertyCallback::invokeIndexedPropertyDeleter(uint32_t index, const v8::PropertyCallbackInfo<v8::Boolean>& info) {
+    v8::Isolate* isolate = info.GetIsolate();
+
+    VALUE code(unwrapDeleter(isolate, info.Data()));
+
+    Unlocker unlock(info.GetIsolate());
+    rb_funcall(code, rb_intern("call"), 2, UINT2NUM(index), (VALUE)PropertyCallbackInfo::Boolean(info));
+  }
+
+  void PropertyCallback::invokeIndexedPropertyEnumerator(const v8::PropertyCallbackInfo<v8::Array>& info) {
+    v8::Isolate* isolate = info.GetIsolate();
+
+    VALUE code(unwrapEnumerator(isolate, info.Data()));
+
+    Unlocker unlock(info.GetIsolate());
+    rb_funcall(code, rb_intern("call"), 1, (VALUE)PropertyCallbackInfo::Array(info));
+  }
+
 }
```

Comment:
Rather than have inline functions to call, I prefer the style used by `FunctionCallback` where it is caste to the different callback types is clearer and involves less repetition. All of the `invoke*` style methods would remain the same, but the difference is that how the pointer to them is achived, basically we'd have the `PropertyCallback` take code and data and then be able to caste to the type of callback needed at the time. E.g.

```c++
// always safe, we never risk passing in a function pointer, when no callback is available
inline operator v8::NamedPropertyGetterCallback() {
  if (RTEST(getter)) {
    return &invokeNamedPropertyGetter;
  } else {
    return 0;
  }
```
Where it's currently calling `wrapData`, we can do that in the cast to `Local<Value>`:

```c++
  inline operator v8::Local<v8::Value>() {
    //wrap the callback and data
  }
```